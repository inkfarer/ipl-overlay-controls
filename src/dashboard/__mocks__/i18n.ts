import i18next from 'i18next';
import { config } from '@vue/test-utils';

jest.mock('i18next-vue', () => ({
    __esModule: true,
    useTranslation: () => ({ t: i18next.t })
}));

beforeAll(async () => {
    await i18next.init({
        lng: 'cimode',
        appendNamespaceToCIMode: true
    });
    config.global.mocks = {
        $t: i18next.t
    };
});
