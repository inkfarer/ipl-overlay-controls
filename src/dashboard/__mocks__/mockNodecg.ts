import Mock = jest.Mock;

export type ReplicantChangeHandler = (newValue?: unknown, oldValue?: unknown) => void;
export type MockReplicant = { value: unknown };
type OnFunction = (event: string, handler: ReplicantChangeHandler) => void;

export class MockNodecg {
    replicants: {[key: string]: MockReplicant};
    listeners: {[key: string]: ReplicantChangeHandler};
    sendMessage: Mock

    constructor() {
        this.replicants = {};
        this.listeners = {};
        this.sendMessage = jest.fn();
    }

    init(): void {
        const self = this;
        window.nodecg = {
            // @ts-ignore: Just a mock
            Replicant(name: string) {
                const replicantValue: { value: unknown, on: OnFunction } = {
                    value: undefined,
                    on: (event: string, handler: ReplicantChangeHandler) => {
                        self.listeners[name] = handler;
                    }
                };
                self.replicants[name] = replicantValue;
                return replicantValue;
            },
            sendMessage: self.sendMessage,
            // @ts-ignore: Just a mock
            log: {
                info: jest.fn()
            }
        };

        // @ts-ignore: It's a mock.
        window.NodeCG = {
            waitForReplicants: jest.fn().mockResolvedValue(true)
        };
    }
}
