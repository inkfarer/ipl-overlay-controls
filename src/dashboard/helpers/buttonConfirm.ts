/*
buttonConfirm

After importing this script to a dashboard panel, add the attribute 'data-requires-confirmation' to any button.
The button will now ask for confirmation when it is first clicked.
When it is clicked a second time, a 'confirm' event is sent by it, which can be listened for by scripts.

Example:

HTML:
    <button id="dangerous-action-button" data-requires-confirmation>DO A THING</button>
JS:
    document.getElementById('dangerous-action-button').addEventListener('confirm', () => {
        // do something
    });
*/

window.addEventListener('DOMContentLoaded', () => {
    function addListener(btn: HTMLButtonElement) {
        if (btn.dataset.hasConfirmListener !== 'true') {
            btn.addEventListener('click', e => {
                const target = e.target as HTMLButtonElement;

                function resetButton() {
                    target.dataset.isClicked = 'false';
                    target.innerText = target.dataset.originalText;
                    target.style.backgroundColor = '';
                }

                if (target.dataset.isClicked === 'true') {
                    target.dispatchEvent(new Event('confirm'));
                    resetButton();
                } else {
                    target.dataset.isClicked = 'true';
                    target.dataset.originalText = target.innerText;
                    target.style.backgroundColor = 'var(--red)';
                    target.innerText = 'are you sure?';

                    setTimeout(() => {
                        resetButton();
                    }, 5000);
                }
            });

            btn.dataset.hasConfirmListener = 'true';
        }
    }

    const buttons = document.querySelectorAll('button[data-requires-confirmation]');
    buttons.forEach(addListener);

    const observer = new MutationObserver(mutations => {
        mutations.forEach(mutation => {
            mutation.addedNodes.forEach(node => {
                if (node.nodeType === Node.ELEMENT_NODE
                    && node.nodeName === 'BUTTON'
                    && (node as HTMLElement).dataset.requiresConfirmation) {

                    const btn = node as HTMLButtonElement;
                    addListener(btn);
                }
            });
        });
    });

    observer.observe(document.body, {
        subtree: true,
        childList: true
    });
});
