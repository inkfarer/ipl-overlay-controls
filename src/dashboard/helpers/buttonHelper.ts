import { colors } from '../styles/colors';

export function setToggleButtonDisabled(
    toggleTrue: HTMLButtonElement,
    toggleFalse: HTMLButtonElement,
    state: boolean
): void {
    toggleTrue.disabled = state;
    toggleFalse.disabled = !state;
}

export function addChangeReminder(
    elements: HTMLElement[] | NodeListOf<HTMLElement>,
    updateButton: HTMLButtonElement
): void {
    elements.forEach(element => {
        if (!element.tagName) return;

        let event;
        if (element.tagName.toLowerCase() === 'input') event = 'input';
        else if (element.tagName.toLowerCase() === 'select') event = 'change';
        else return;

        element.addEventListener(event, e => {
            const target = e.target as HTMLElement;
            target.dataset.edited = 'true';
            updateButton.style.backgroundColor = colors.red;
        });
    });

    updateButton.addEventListener('click', () => {
        elements.forEach(elem => {
            elem.dataset.edited = 'false';
        });
        updateButton.style.backgroundColor = colors.blue;
    });
}
