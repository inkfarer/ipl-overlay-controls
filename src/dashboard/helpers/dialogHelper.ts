import { NodecgDialog } from '../types/dialog';

export function closeDialog(dialogName: string): void {
    (nodecg.getDialog(dialogName) as NodecgDialog).close();
}
