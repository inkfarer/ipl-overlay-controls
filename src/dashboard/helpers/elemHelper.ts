export function appendChildren(element: HTMLElement, ...children: HTMLElement[]): void {
    children.forEach(elem => {
        element.appendChild(elem);
    });
}

export function elementById<T extends HTMLElement>(id: string): T | undefined {
    return document.getElementById(id) as T;
}

export function dispatch(element: EventTarget, event: string): void {
    element.dispatchEvent(new Event(event));
}

export function hideElement(element: HTMLElement): void {
    element.style.display = 'none';
}

export function showElement(element: HTMLElement): void {
    element.style.removeProperty('display');
}

export function prependToBody(elem: HTMLElement): void {
    document.body.prepend(elem);
}

export function appendElementAfter(newElement: HTMLElement, element: HTMLElement): void {
    element.parentNode.insertBefore(newElement, element.nextSibling);
}
