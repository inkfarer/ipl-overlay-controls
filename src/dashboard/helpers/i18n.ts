import { RuntimeConfig } from 'schemas';
import i18next, { type BackendModule } from 'i18next';

const I18nBackend: BackendModule = {
    type: 'backend',

    // eslint-disable-next-line @typescript-eslint/no-empty-function
    init() {

    },

    async read(language, namespace) {
        return loadTranslation(language, namespace);
    },
};

export async function initI18n(namespaces: string | string[]): Promise<void> {
    const ns = Array.isArray(namespaces) ? [...namespaces, 'common']: [namespaces, 'common'];
    await i18next
        .use(I18nBackend)
        .init({
            lng: 'en',
            fallbackLng: 'en',
            ns,
            interpolation: {
                escapeValue: false
            }
        });

    const runtimeConfig = nodecg.Replicant<RuntimeConfig>('runtimeConfig');
    runtimeConfig.on('change', async (newValue, oldValue) => {
        if (!oldValue || newValue.interfaceLocale !== oldValue.interfaceLocale) {
            await i18next.changeLanguage(newValue.interfaceLocale.toLowerCase()).catch(e => {
                console.error('Failed to change interface language', e);
            });
            if (i18next.exists('title')) {
                try {
                    window.frameElement.parentElement.setAttribute('display-title', i18next.t('title'));
                } catch (ignore) {

                }
            }
        }
    });
}

async function loadTranslation(locale: string, name: string) {
    try {
        return await import(`../../helpers/i18n/${locale.toLowerCase()}/${name}.json`);
    } catch (ignore) {
        return null;
    }
}
