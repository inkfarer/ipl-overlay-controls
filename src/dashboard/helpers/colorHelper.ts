import { colors, getColorOptionName } from '../../helpers/splatoonData';

export function fillColorSelector(select: HTMLSelectElement): void {
    for (let i = 0; i < colors.length; i++) {
        const element = colors[i];

        const optGroup = document.createElement('optgroup');
        optGroup.label = element.meta.name;

        for (let j = 0; j < element.colors.length; j++) {
            const color = element.colors[j];

            const option = document.createElement('option');
            option.value = getColorOptionName(color.index, element.meta.name);
            option.text = color.title;
            option.dataset.index = String(color.index);
            option.dataset.firstColor = color.clrA;
            option.dataset.secondColor = color.clrB;
            option.dataset.categoryName = element.meta.name;
            option.disabled = color.isCustom;

            optGroup.appendChild(option);
        }

        select.appendChild(optGroup);
    }
}

export function getContrastingTextColor(backgroundColor: string): string {
    const { r, g, b } = hexToRbg(backgroundColor);
    const yiq = ((r * 299) + (g * 587) + (b * 114)) / 1000;
    return (yiq >= 128) ? '#333' : 'white';
}

export function hexToRbg(color: string): { r: number, g: number, b: number } {
    color = color.replace('#', '');
    return {
        r: parseInt(color.substr(0, 2), 16),
        g: parseInt(color.substr(2, 2), 16),
        b: parseInt(color.substr(4, 2), 16)
    };
}
