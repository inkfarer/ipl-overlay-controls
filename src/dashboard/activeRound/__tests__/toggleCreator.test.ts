import { MockNodecg } from '../../__mocks__/mockNodecg';
import { Module } from '../../../helpers/__mocks__/module';
import { dispatch, elementById } from '../../helpers/elemHelper';
import { GameWinner } from 'types/enums/gameWinner';

describe('toggleCreator', () => {
    const mockToggleHelper = {
        __esModule: true,
        getActiveRoundColors: jest.fn(),
        handleColorSelectorChange: jest.fn(),
        handleColorSwapToggleChange: jest.fn(),
        handleCustomColorListenerChange: jest.fn(),
        setToggleColor: jest.fn(),
        toggleCustomColorSelectorVisibility: jest.fn()
    };
    let nodecg: MockNodecg;
    let helper: Module;

    jest.mock('../main', () => ({
        __esModule: true,
        enableColorEditToggle: document.createElement('input')
    }));
    jest.mock('../toggleHelper', () => mockToggleHelper);

    beforeEach(() => {
        jest.resetModules();
        jest.resetAllMocks();

        nodecg = new MockNodecg();
        nodecg.init();

        document.body.innerHTML = `
            <input id="enable-color-edit-toggle">
            <button id="update-round">
            <div id="toggles" />`;

        helper = require('../toggleCreator');
    });

    describe('addToggle', () => {
        const setupWithColorData = () => {
            helper.addToggle({
                color: {
                    clrA: '#123',
                    clrB: '#aaa',
                    index: 1,
                    categoryName: 'Cool Color',
                    colorsSwapped: false
                },
                stage: 'Blackbelly Skatepark',
                mode: 'Splat Zones'
            }, 0);
        };

        it('creates toggles when the round has color data', () => {
            helper.addToggle({
                color: {
                    clrA: '#123',
                    clrB: '#aaa',
                    index: 1,
                    categoryName: 'Cool Color',
                    colorsSwapped: false
                },
                stage: 'Blackbelly Skatepark',
                mode: 'Splat Zones'
            }, 0);

            expect(elementById('toggles').innerHTML).toMatchSnapshot();
        });

        it('creates toggles when the round has no color data', () => {
            mockToggleHelper.getActiveRoundColors.mockReturnValue({
                activeColor: {
                    index: 2,
                    categoryName: 'Cool Color (Active)'
                },
                teamA: { color: '#456' },
                teamB: { color: '#EEE' }
            });

            helper.addToggle({
                color: undefined,
                stage: 'MakoMart',
                mode: 'Rainmaker'
            }, 0);

            expect(elementById('toggles').innerHTML).toMatchSnapshot();
        });

        it('creates event listeners for custom color selectors', () => {
            setupWithColorData();

            dispatch(elementById('custom-color-selector_a_0'), 'change');
            dispatch(elementById('custom-color-selector_b_0'), 'change');

            expect(mockToggleHelper.handleCustomColorListenerChange).toHaveBeenCalledTimes(2);
        });

        it('creates event listeners for color selectors', () => {
            setupWithColorData();

            dispatch(elementById('color-selector_0'), 'change');

            expect(mockToggleHelper.handleColorSelectorChange).toHaveBeenCalled();
        });

        it('creates event listeners for color swap toggles', () => {
            setupWithColorData();

            dispatch(elementById('color-swap-toggle_0'), 'change');

            expect(mockToggleHelper.handleColorSwapToggleChange).toHaveBeenCalled();
        });

        it('creates event listeners for custom color toggles', () => {
            setupWithColorData();

            dispatch(elementById('custom-color-toggle_0'), 'change');

            expect(mockToggleHelper.toggleCustomColorSelectorVisibility).toHaveBeenCalled();
        });

        it('creates event listeners for win buttons', () => {
            setupWithColorData();

            dispatch(elementById('no-win-toggle_0'), 'click');
            dispatch(elementById('team-a-win-toggle_0'), 'click');
            dispatch(elementById('team-b-win-toggle_0'), 'click');

            expect(nodecg.sendMessage).toHaveBeenNthCalledWith(1, 'setWinner',
                { winner: GameWinner.NO_WINNER, roundIndex: 0 });
            expect(nodecg.sendMessage).toHaveBeenNthCalledWith(2, 'setWinner',
                { winner: GameWinner.ALPHA, roundIndex: 0 });
            expect(nodecg.sendMessage).toHaveBeenNthCalledWith(3, 'setWinner',
                { winner: GameWinner.BRAVO, roundIndex: 0 });
        });
    });
});
