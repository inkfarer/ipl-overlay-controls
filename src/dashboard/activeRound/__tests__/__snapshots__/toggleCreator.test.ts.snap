// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toggleCreator addToggle creates toggles when the round has color data 1`] = `"<div class=\\"toggles layout horizontal center-horizontal even-spacing\\" id=\\"game-editor_0\\"><div class=\\"round-number-display\\"></div><select class=\\"stage-selector\\" id=\\"stage-selector_0\\"><option value=\\"Ancho-V Games\\">Ancho-V Games</option><option value=\\"Arowana Mall\\">Arowana Mall</option><option value=\\"Blackbelly Skatepark\\">Blackbelly Skatepark</option><option value=\\"Camp Triggerfish\\">Camp Triggerfish</option><option value=\\"Goby Arena\\">Goby Arena</option><option value=\\"Humpback Pump Track\\">Humpback Pump Track</option><option value=\\"Inkblot Art Academy\\">Inkblot Art Academy</option><option value=\\"Kelp Dome\\">Kelp Dome</option><option value=\\"MakoMart\\">MakoMart</option><option value=\\"Manta Maria\\">Manta Maria</option><option value=\\"Moray Towers\\">Moray Towers</option><option value=\\"Musselforge Fitness\\">Musselforge Fitness</option><option value=\\"New Albacore Hotel\\">New Albacore Hotel</option><option value=\\"Piranha Pit\\">Piranha Pit</option><option value=\\"Port Mackerel\\">Port Mackerel</option><option value=\\"Shellendorf Institute\\">Shellendorf Institute</option><option value=\\"Shifty Station\\">Shifty Station</option><option value=\\"Skipper Pavilion\\">Skipper Pavilion</option><option value=\\"Snapper Canal\\">Snapper Canal</option><option value=\\"Starfish Mainstage\\">Starfish Mainstage</option><option value=\\"Sturgeon Shipyard\\">Sturgeon Shipyard</option><option value=\\"The Reef\\">The Reef</option><option value=\\"Unknown Stage\\">Unknown Stage</option><option value=\\"Wahoo World\\">Wahoo World</option><option value=\\"Walleye Warehouse\\">Walleye Warehouse</option></select><select class=\\"mode-selector\\" id=\\"mode-selector_0\\"><option value=\\"Clam Blitz\\">Clam Blitz</option><option value=\\"Rainmaker\\">Rainmaker</option><option value=\\"Splat Zones\\">Splat Zones</option><option value=\\"Tower Control\\">Tower Control</option><option value=\\"Turf War\\">Turf War</option><option value=\\"Unknown Mode\\">Unknown Mode</option></select><div style=\\"display: none;\\" class=\\"color-selector-wrapper layout horizontal center-vertical even-spacing\\"><select class=\\"color-selector\\" id=\\"color-selector_0\\" data-source=\\"gameInfo\\"><optgroup label=\\"Ranked Modes\\"><option value=\\"ranked-modes_0\\" data-index=\\"0\\" data-first-color=\\"#37FC00\\" data-second-color=\\"#7D28FC\\" data-category-name=\\"Ranked Modes\\">Green vs Grape</option><option value=\\"ranked-modes_1\\" data-index=\\"1\\" data-first-color=\\"#04D976\\" data-second-color=\\"#D600AB\\" data-category-name=\\"Ranked Modes\\">Green vs Magenta</option><option value=\\"ranked-modes_2\\" data-index=\\"2\\" data-first-color=\\"#10E38F\\" data-second-color=\\"#FB7B08\\" data-category-name=\\"Ranked Modes\\">Turquoise vs Orange</option><option value=\\"ranked-modes_3\\" data-index=\\"3\\" data-first-color=\\"#FF9E03\\" data-second-color=\\"#B909E0\\" data-category-name=\\"Ranked Modes\\">Mustard vs Purple</option><option value=\\"ranked-modes_4\\" data-index=\\"4\\" data-first-color=\\"#2F27CC\\" data-second-color=\\"#37FC00\\" data-category-name=\\"Ranked Modes\\">Dark Blue vs Green</option><option value=\\"ranked-modes_5\\" data-index=\\"5\\" data-first-color=\\"#B909E0\\" data-second-color=\\"#37FC00\\" data-category-name=\\"Ranked Modes\\">Purple vs Green</option><option value=\\"ranked-modes_6\\" data-index=\\"6\\" data-first-color=\\"#FEF232\\" data-second-color=\\"#2ED2FE\\" data-category-name=\\"Ranked Modes\\">Yellow vs Blue</option></optgroup><optgroup label=\\"Turf War\\"><option value=\\"turf-war_0\\" data-index=\\"0\\" data-first-color=\\"#D1E004\\" data-second-color=\\"#960CB2\\" data-category-name=\\"Turf War\\">Yellow vs Purple</option><option value=\\"turf-war_1\\" data-index=\\"1\\" data-first-color=\\"#E61077\\" data-second-color=\\"#361CB8\\" data-category-name=\\"Turf War\\">Pink vs Blue</option><option value=\\"turf-war_2\\" data-index=\\"2\\" data-first-color=\\"#ED0C6A\\" data-second-color=\\"#D5E802\\" data-category-name=\\"Turf War\\">Pink vs Yellow</option><option value=\\"turf-war_3\\" data-index=\\"3\\" data-first-color=\\"#6B10CC\\" data-second-color=\\"#08CC81\\" data-category-name=\\"Turf War\\">Purple vs Turquoise</option><option value=\\"turf-war_4\\" data-index=\\"4\\" data-first-color=\\"#E30960\\" data-second-color=\\"#02ADCF\\" data-category-name=\\"Turf War\\">Pink vs Light Blue</option><option value=\\"turf-war_5\\" data-index=\\"5\\" data-first-color=\\"#5617C2\\" data-second-color=\\"#FF5F03\\" data-category-name=\\"Turf War\\">Purple vs Orange</option><option value=\\"turf-war_6\\" data-index=\\"6\\" data-first-color=\\"#E60572\\" data-second-color=\\"#1BBF0F\\" data-category-name=\\"Turf War\\">Pink vs Green</option></optgroup><optgroup label=\\"Color Lock\\"><option value=\\"color-lock_0\\" data-index=\\"0\\" data-first-color=\\"#FEF232\\" data-second-color=\\"#2F27CC\\" data-category-name=\\"Color Lock\\">Yellow vs Blue (Color Lock)</option></optgroup><optgroup label=\\"Custom Color\\"><option value=\\"custom-color_0\\" data-index=\\"0\\" data-first-color=\\"#000000\\" data-second-color=\\"#FFFFFF\\" data-category-name=\\"Custom Color\\" disabled=\\"\\">Custom Color</option></optgroup></select><div class=\\"custom-color-select-wrapper layout horizontal\\" id=\\"custom-color-select-wrapper_0\\" style=\\"display: none;\\"><input type=\\"color\\" id=\\"custom-color-selector_a_0\\" class=\\"team-a\\"><input type=\\"color\\" id=\\"custom-color-selector_b_0\\" class=\\"team-b\\"></div><div class=\\"layout horizontal center-horizontal color-swap-toggle-container\\"><input class=\\"color-swap-toggle\\" id=\\"color-swap-toggle_0\\" type=\\"checkbox\\" data-source=\\"gameInfo\\"><label for=\\"color-swap-toggle_0\\" class=\\"white-label\\"></label><input class=\\"custom-color-toggle\\" id=\\"custom-color-toggle_0\\" type=\\"checkbox\\"><label class=\\"white-label\\" for=\\"custom-color-toggle_0\\"></label></div></div><div class=\\"layout horizontal win-button-container\\"><button class=\\"no-win-toggle max-width red\\" id=\\"no-win-toggle_0\\"><i class=\\"fas fa-times\\"></i></button><button class=\\"team-a-win-toggle max-width win-set-toggle\\" style=\\"background-color: rgb(17, 34, 51); color: white; border-left: 0 solid #123;\\" id=\\"team-a-win-toggle_0\\">A</button><button class=\\"team-b-win-toggle max-width win-set-toggle\\" style=\\"background-color: rgb(170, 170, 170); color: white; border-right: 0 solid #aaa;\\" id=\\"team-b-win-toggle_0\\">B</button></div></div>"`;

exports[`toggleCreator addToggle creates toggles when the round has no color data 1`] = `"<div class=\\"toggles layout horizontal center-horizontal even-spacing\\" id=\\"game-editor_0\\"><div class=\\"round-number-display\\"></div><select class=\\"stage-selector\\" id=\\"stage-selector_0\\"><option value=\\"Ancho-V Games\\">Ancho-V Games</option><option value=\\"Arowana Mall\\">Arowana Mall</option><option value=\\"Blackbelly Skatepark\\">Blackbelly Skatepark</option><option value=\\"Camp Triggerfish\\">Camp Triggerfish</option><option value=\\"Goby Arena\\">Goby Arena</option><option value=\\"Humpback Pump Track\\">Humpback Pump Track</option><option value=\\"Inkblot Art Academy\\">Inkblot Art Academy</option><option value=\\"Kelp Dome\\">Kelp Dome</option><option value=\\"MakoMart\\">MakoMart</option><option value=\\"Manta Maria\\">Manta Maria</option><option value=\\"Moray Towers\\">Moray Towers</option><option value=\\"Musselforge Fitness\\">Musselforge Fitness</option><option value=\\"New Albacore Hotel\\">New Albacore Hotel</option><option value=\\"Piranha Pit\\">Piranha Pit</option><option value=\\"Port Mackerel\\">Port Mackerel</option><option value=\\"Shellendorf Institute\\">Shellendorf Institute</option><option value=\\"Shifty Station\\">Shifty Station</option><option value=\\"Skipper Pavilion\\">Skipper Pavilion</option><option value=\\"Snapper Canal\\">Snapper Canal</option><option value=\\"Starfish Mainstage\\">Starfish Mainstage</option><option value=\\"Sturgeon Shipyard\\">Sturgeon Shipyard</option><option value=\\"The Reef\\">The Reef</option><option value=\\"Unknown Stage\\">Unknown Stage</option><option value=\\"Wahoo World\\">Wahoo World</option><option value=\\"Walleye Warehouse\\">Walleye Warehouse</option></select><select class=\\"mode-selector\\" id=\\"mode-selector_0\\"><option value=\\"Clam Blitz\\">Clam Blitz</option><option value=\\"Rainmaker\\">Rainmaker</option><option value=\\"Splat Zones\\">Splat Zones</option><option value=\\"Tower Control\\">Tower Control</option><option value=\\"Turf War\\">Turf War</option><option value=\\"Unknown Mode\\">Unknown Mode</option></select><div style=\\"display: none;\\" class=\\"color-selector-wrapper layout horizontal center-vertical even-spacing\\"><select class=\\"color-selector\\" id=\\"color-selector_0\\" data-source=\\"scoreboard\\"><optgroup label=\\"Ranked Modes\\"><option value=\\"ranked-modes_0\\" data-index=\\"0\\" data-first-color=\\"#37FC00\\" data-second-color=\\"#7D28FC\\" data-category-name=\\"Ranked Modes\\">Green vs Grape</option><option value=\\"ranked-modes_1\\" data-index=\\"1\\" data-first-color=\\"#04D976\\" data-second-color=\\"#D600AB\\" data-category-name=\\"Ranked Modes\\">Green vs Magenta</option><option value=\\"ranked-modes_2\\" data-index=\\"2\\" data-first-color=\\"#10E38F\\" data-second-color=\\"#FB7B08\\" data-category-name=\\"Ranked Modes\\">Turquoise vs Orange</option><option value=\\"ranked-modes_3\\" data-index=\\"3\\" data-first-color=\\"#FF9E03\\" data-second-color=\\"#B909E0\\" data-category-name=\\"Ranked Modes\\">Mustard vs Purple</option><option value=\\"ranked-modes_4\\" data-index=\\"4\\" data-first-color=\\"#2F27CC\\" data-second-color=\\"#37FC00\\" data-category-name=\\"Ranked Modes\\">Dark Blue vs Green</option><option value=\\"ranked-modes_5\\" data-index=\\"5\\" data-first-color=\\"#B909E0\\" data-second-color=\\"#37FC00\\" data-category-name=\\"Ranked Modes\\">Purple vs Green</option><option value=\\"ranked-modes_6\\" data-index=\\"6\\" data-first-color=\\"#FEF232\\" data-second-color=\\"#2ED2FE\\" data-category-name=\\"Ranked Modes\\">Yellow vs Blue</option></optgroup><optgroup label=\\"Turf War\\"><option value=\\"turf-war_0\\" data-index=\\"0\\" data-first-color=\\"#D1E004\\" data-second-color=\\"#960CB2\\" data-category-name=\\"Turf War\\">Yellow vs Purple</option><option value=\\"turf-war_1\\" data-index=\\"1\\" data-first-color=\\"#E61077\\" data-second-color=\\"#361CB8\\" data-category-name=\\"Turf War\\">Pink vs Blue</option><option value=\\"turf-war_2\\" data-index=\\"2\\" data-first-color=\\"#ED0C6A\\" data-second-color=\\"#D5E802\\" data-category-name=\\"Turf War\\">Pink vs Yellow</option><option value=\\"turf-war_3\\" data-index=\\"3\\" data-first-color=\\"#6B10CC\\" data-second-color=\\"#08CC81\\" data-category-name=\\"Turf War\\">Purple vs Turquoise</option><option value=\\"turf-war_4\\" data-index=\\"4\\" data-first-color=\\"#E30960\\" data-second-color=\\"#02ADCF\\" data-category-name=\\"Turf War\\">Pink vs Light Blue</option><option value=\\"turf-war_5\\" data-index=\\"5\\" data-first-color=\\"#5617C2\\" data-second-color=\\"#FF5F03\\" data-category-name=\\"Turf War\\">Purple vs Orange</option><option value=\\"turf-war_6\\" data-index=\\"6\\" data-first-color=\\"#E60572\\" data-second-color=\\"#1BBF0F\\" data-category-name=\\"Turf War\\">Pink vs Green</option></optgroup><optgroup label=\\"Color Lock\\"><option value=\\"color-lock_0\\" data-index=\\"0\\" data-first-color=\\"#FEF232\\" data-second-color=\\"#2F27CC\\" data-category-name=\\"Color Lock\\">Yellow vs Blue (Color Lock)</option></optgroup><optgroup label=\\"Custom Color\\"><option value=\\"custom-color_0\\" data-index=\\"0\\" data-first-color=\\"#000000\\" data-second-color=\\"#FFFFFF\\" data-category-name=\\"Custom Color\\" disabled=\\"\\">Custom Color</option></optgroup></select><div class=\\"custom-color-select-wrapper layout horizontal\\" id=\\"custom-color-select-wrapper_0\\" style=\\"display: none;\\"><input type=\\"color\\" id=\\"custom-color-selector_a_0\\" class=\\"team-a\\"><input type=\\"color\\" id=\\"custom-color-selector_b_0\\" class=\\"team-b\\"></div><div class=\\"layout horizontal center-horizontal color-swap-toggle-container\\"><input class=\\"color-swap-toggle\\" id=\\"color-swap-toggle_0\\" type=\\"checkbox\\" data-source=\\"scoreboard\\"><label for=\\"color-swap-toggle_0\\" class=\\"white-label\\"></label><input class=\\"custom-color-toggle\\" id=\\"custom-color-toggle_0\\" type=\\"checkbox\\"><label class=\\"white-label\\" for=\\"custom-color-toggle_0\\"></label></div></div><div class=\\"layout horizontal win-button-container\\"><button class=\\"no-win-toggle max-width red\\" id=\\"no-win-toggle_0\\"><i class=\\"fas fa-times\\"></i></button><button class=\\"team-a-win-toggle max-width win-set-toggle\\" id=\\"team-a-win-toggle_0\\">A</button><button class=\\"team-b-win-toggle max-width win-set-toggle\\" id=\\"team-b-win-toggle_0\\">B</button></div></div>"`;
